### ISSUE TITLE
SIMD Numeric Arrays Operations

### ISSUE THREAD

**Issue by MohammadHamdyGhanem** [dotnet/corefx#26251]
I suggest to add some SIMD operations to Array class, such as:

```
int[] a = {1, 2, 3, 4, 5, 6, 7, 8};
int[] b = {1, 2, 3, 4, 5, 6, 7, 8};
var c = a + b ;
```

C# should contain {2, 4, 6, 8, 10, 12, 14, 16}
If a and b have different  lengths, C# can raise exception, or conceder missed elements in the shortest array to be zeroes.
If we want to do so, we can use Vector4 to calculate the result:
var Result1 = new Vector4(1, 2, 3, 4) + new Vector4(1, 2, 3, 4);
var Result2 = new Vector4(5, 6, 7, 8) + new Vector4(5, 6, 7, 8);
Then add the elements from the two results to the array. This will be verbose and ugly long code to write! It's better to write this in a low level SIMD directly without the need for Vector4.
Likewise, you can provide these operations for numeric arrays:
```
c = a - b ;
c = a * b ;
c = a / b ;
```
Also, b can a scalar value that operates on all array elements:
```
c = a + 1 ;
c = a - 1 ;
c = a * 2 ;
c = a / 3 ;
```

Maybe it's better to add a class Named Simd to the System.Numerics namespace, that performs these operations on numeric arrays.

```
c = Simd.Add (a, b)
c = Simd.Add (a, 1)
```
The same for Simd.Subtract , Simd.Multiply, Simd.Divide, Simd.Pow, Simd.Aggregate, etc.
Add as many common array tasks as you can, that can be use the SIMD feature.

**Comment by tannergooding** [comment]
See the Hardware Intrinsics proposal: https://github.com/dotnet/designs/blob/master/accepted/platform-intrinsics.md (doc is slightly out of date, based on the most recent work).

**Comment by aobatact** [comment]
Current solution might be like this
```csharp
public static int[] Add(this int[] x,int[] y)
{
    if (x.Length != y.Length) throw new ArgumentException();
    var result = new int[x.Length];
    int i = 0;
    for (;i < x.Length; i+= Vector<int>.Count)
    {
        (new Vector<int>(x, i) + new Vector<int>(y, i)).CopyTo(result, i);
    }
    for (; i < x.Length; i++)
    {
        result[i] = x[i] + y[i];
    }
    return result;
}
public static int[] Add(this int[] x, int y)
{
    var result = new int[x.Length];
    int i = 0;
    for (; i < x.Length; i += Vector<int>.Count)
    {
        (new Vector<int>(x, i) + new Vector<int>(y)).CopyTo(result, i);
    }
    for (; i < x.Length; i++)
    {
        result[i] = x[i] + y;
    }
    return result;
}
```

**Comment by MohammadHamdyGhanem** [comment]
@aobatact 
Yes. This is the scary code that makes my proposal in first place! .. Vector 4 seems great, but practically it consumes a lot of coding effort and I think the way you fill it with values from arrays and get result back to arrays will eat some of the saved execution time! So, the question is: why don't you do this in a more efficient low level way, and make us happy?

**Comment by tannergooding** [comment]
> I think the way you fill it with values from arrays and get result back to arrays will eat some of the saved execution time

@MohammadHamdyGhanem, have you checked the codegen to see if Vector4 is actually not performing well?

The managed API may look one way, but the actual codegen tends to be fairly efficient. There is also the `Unsafe.CopyBlock`, `Unsafe.Read`, and `Unsafe.Write` APIs (in the System.Runtime.CompilerServices namespace) that let you efficiently read/write `Vector<T>` from/to memory.

As for doing it a more low level way (directly executing SIMD instructions), that scenario is covered by the Hardware Intrinsics feature I linked above.

**Comment by MohammadHamdyGhanem** [comment]
@tannergooding 
Good. So, put them all togather and give us an easy to use class or extension methods to the Array class.

**Comment by MohammadHamdyGhanem** [comment]
@tannergooding 
Is there a plan to use Span<T> with Vectors?
It can be faster then copying array elements to the Vector.
Also, the static Vector calss can use it to do operations on two arrays in slices. Ecah method in the Vector class can have another overload with two T[] params instead of Vector[T] param. 
What do you think?

**Comment by svick** [comment]
Would using SIMD even make sense here? These operations allocate new arrays, and I think that for non-trivial applications, the increased time spent in GC from that will be significantly higher than the improvement from using SIMD.

Though that point would be moot if the API did not allocate arrays, probably by using `Span`s.

**Comment by MohammadHamdyGhanem** [comment]
If your code already requires doing some operations on two large arrays and get the result in a third one, the cost of creating arrays is fixed with or without SIMD, but using SIMD can save some time.
Also, There is no built in array functions in framework, it is nice to have some.

**Comment by MohammadHamdyGhanem** [comment]
W about the return value be a Memory< T > ?

**Comment by tannergooding** [comment]
> What about to make the return value a Memory< T > ?

@MohammadHamdyGhanem, how would that help?

The benefit of using `Vector<T>` is that it uses SIMD (Single Instruction, Multiple Data) instructions (on supported hardware). 

For loading/storing operations, having things like `Span<T>` is beneficial because it lets you read/write part way into a large array (without having to pass around the array plus the start index as separate arguments).

However, returning anything except for `Vector<T>` itself is less efficient, as that means you lose the ability for it to be kept in register and instead get an indirection to the actual value (which will always incur a write/etc)

**Comment by MohammadHamdyGhanem** [comment]
@tannergooding 
I speak about array operations sliced internally to be done by vectors. The return value is an array,  a span< T > or a memory < T >. This will shorten the code when doing operation on arrays.

**Comment by MohammadHamdyGhanem** [comment]
@tannergooding 
I have a question about performance:
Suppose we have  an array with 100 integer elements, and we need to compute their average. what is the fastest way of these:
-	A traditional summation loop
-	LinQ aggregation
-	putting the array elements in 25 Vector< int >, and write a recursive loop to sum them in pairs until we end up with one vector that we use to calculate the final summation. 
Can using vectors like this save any time?.. Supposing that vector < T> operates on the ref of the original arrays and we don't copy its values. I know that recursive functions will add an overhead, but I can't think of another way that makes use of the SIMD with these 100 Addition operations .
Of course in real statistics, there can thousands of element! Is there a way to reduse the time of these repeated operations?

**Comment by MohammadHamdyGhanem** [comment]
I used Vectors like this (without recursion) 
```
    Function SumArray(A() As Integer) As Integer
        Dim Result As New Vector(Of Integer)
        For i = 0 To A.Length - 1 Step 4
            Result += New Vector(Of Integer)(A, i)
        Next
        Return Result(0) + Result(1) + Result(2) + Result(3)
    End Function

```
And did some testing.. These are the results:
Traditional For Loop: 4.5 ms
Vectors with no hardware acceleration: 33ms
Vectors with hardware acceleration: 2.5 ms
LinQ: 10ms
LinQ (AsParallel): The first call gives very bad results, but next calls takes 3.5ms in average.

Notes:
1. By default the project is configured to run on "Any CPU". Vectors are not accelerated with this config. even if the hardware is capable and operating system is win 10 64bit! Is it possible to change this behaviour?

2. Wouldn't be nice if there is a Vector.Sum instead of using Result(0) + Result(1) + Result(2) + Result(3)?
You can add som aggregation methods to Vectors, or at least let vectors implement IReadOnlyCollection, so we can use it with for each and IinQ.

3. When I was testing this LinQ:
Dim s = Aggregate x In Arr Into Sum() 
It gave me out of memory exception. It is strange to make the retutn value of the Sum function of the same type as the elements of the collection! One should expect that aggregating integers can give a long result! How can this be solved?

4. Why does The first call of AsParallel LinQ give very bad results? One expects to do the aggregation only one tine, and this will talk longer time than non parallel LinQ!
